cpp/src/arrow/builder.cc

```diff
@@ -81,35 +81,29 @@ Status ArrayBuilder::AppendToBitmap(const uint8_t* valid_bytes, int64_t length)
   return Status::OK();
 }
 
-Status ArrayBuilder::Init(int64_t capacity) {
-  int64_t to_alloc = BitUtil::BytesForBits(capacity);
-  null_bitmap_ = std::make_shared<PoolBuffer>(pool_);
-  RETURN_NOT_OK(null_bitmap_->Resize(to_alloc));
-
-  // Buffers might allocate more then necessary to satisfy padding requirements
-  const int64_t byte_capacity = null_bitmap_->capacity();
-  capacity_ = capacity;
-  null_bitmap_data_ = null_bitmap_->mutable_data();
-  memset(null_bitmap_data_, 0, static_cast<size_t>(byte_capacity));
-
-  return Status::OK();
-}
-
-Status ArrayBuilder::Resize(int64_t new_bits) {
+Status ArrayBuilder::Resize(int64_t capacity) {
   if (!null_bitmap_) {
-    return Init(new_bits);
-  }
-  int64_t new_bytes = BitUtil::BytesForBits(new_bits);
-  int64_t old_bytes = null_bitmap_->size();
-  RETURN_NOT_OK(null_bitmap_->Resize(new_bytes));
-  null_bitmap_data_ = null_bitmap_->mutable_data();
-  // The buffer might be overpadded to deal with padding according to the spec
-  const int64_t byte_capacity = null_bitmap_->capacity();
-  capacity_ = new_bits;
-  if (old_bytes < new_bytes) {
-    memset(null_bitmap_data_ + old_bytes, 0,
-           static_cast<size_t>(byte_capacity - old_bytes));
+    int64_t to_alloc = BitUtil::BytesForBits(capacity);
+    null_bitmap_ = std::make_shared<PoolBuffer>(pool_);
```

Comment:
Maybe use the `AllocateResizableBuffer` here? It will take care of zeroing, too.

Reply:
Aren't we supposed to make `PoolBuffer` a private class? I agree with using `AllocateResizableBuffer`, and zeroing up-to-size.

(also ideally we don't have to zero up-to-size; is our logic for filling up the null bitmap defective?)

Reply:
About the zeroing: we either have to zero in appending bits, which is a little bit complicated logic, of we just zero upfront. I think there is comment about it in old BooleanBuilder::Init / new BooleanBuilder::Resize.

Reply:
I think that depends on the API which is being used. `FirstTimeBitmapWriter` and `GenerateBits` should be immune to Valgrind issues, but `SetBit` and friends are not.

Reply:
Right, might be worth it to switch away from `SetBit` to `FirstTimeBitmapWriter`.

Reply:
I opened https://issues.apache.org/jira/browse/ARROW-2904. I think that change is a bit much for this PR